File: cpp/src/vulkan/swapchain.cpp
#include "vulkan/swapchain.hpp"

#include <algorithm>
#include <stdexcept>

#include "vulkan/vk_check.hpp"

SwapchainContext::~SwapchainContext() {
    // destroy() 应由调用方在 device 销毁前显式调用
}

void SwapchainContext::create(VkInstance instance, VulkanDevice& dev, VkPhysicalDevice pd,
                              VkSurfaceKHR surface, uint32_t width, uint32_t height) {
    device_ = dev.device();
    vkGetDeviceProcAddr(device_, "vkDestroySwapchainKHR",
                        reinterpret_cast<PFN_vkVoidFunction*>(&destroy_swapchain_));
    vkGetDeviceProcAddr(device_, "vkAcquireNextImageKHR",
                        reinterpret_cast<PFN_vkVoidFunction*>(&acquire_next_image_));
    vkGetDeviceProcAddr(device_, "vkQueuePresentKHR",
                        reinterpret_cast<PFN_vkVoidFunction*>(&queue_present_));

    VkSurfaceCapabilitiesKHR caps{};
    vkGetPhysicalDeviceSurfaceCapabilitiesKHR(pd, surface, &caps);

    uint32_t fmt_count = 0;
    vkGetPhysicalDeviceSurfaceFormatsKHR(pd, surface, &fmt_count, nullptr);
    std::vector<VkSurfaceFormatKHR> formats(fmt_count);
    vkGetPhysicalDeviceSurfaceFormatsKHR(pd, surface, &fmt_count, formats.data());

    format_ = formats[0].format;
    VkColorSpaceKHR color_space = formats[0].colorSpace;
    for (const auto& f : formats) {
        if (f.format == VK_FORMAT_B8G8R8A8_SRGB &&
            f.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR) {
            format_ = f.format;
            color_space = f.colorSpace;
            break;
        }
    }

    extent_.width =
        std::clamp(width, caps.minImageExtent.width, caps.maxImageExtent.width);
    extent_.height =
        std::clamp(height, caps.minImageExtent.height, caps.maxImageExtent.height);

    uint32_t image_count = caps.minImageCount + 1;
    if (caps.maxImageCount > 0 && image_count > caps.maxImageCount) {
        image_count = caps.maxImageCount;
    }

    VkSwapchainCreateInfoKHR sci{};
    sci.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
    sci.surface = surface;
    sci.minImageCount = image_count;
    sci.imageFormat = format_;
    sci.imageColorSpace = color_space;
    sci.imageExtent = extent_;
    sci.imageArrayLayers = 1;
    sci.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
    sci.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
    sci.preTransform = caps.currentTransform;
    sci.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
    sci.presentMode = VK_PRESENT_MODE_FIFO_KHR;
    sci.clipped = VK_TRUE;
    sci.oldSwapchain = VK_NULL_HANDLE;

    vk_check(vkCreateSwapchainKHR(device_, &sci, nullptr, &swapchain_), "create swapchain");

    uint32_t sc_count = 0;
    vkGetSwapchainImagesKHR(device_, swapchain_, &sc_count, nullptr);
    images_.resize(sc_count);
    vkGetSwapchainImagesKHR(device_, swapchain_, &sc_count, images_.data());

    image_views_.resize(images_.size());
    for (std::size_t i = 0; i < images_.size(); ++i) {
        VkImageViewCreateInfo vi{};
        vi.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
        vi.image = images_[i];
        vi.viewType = VK_IMAGE_VIEW_TYPE_2D;
        vi.format = format_;
        vi.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
        vi.subresourceRange.levelCount = 1;
        vi.subresourceRange.layerCount = 1;
        vk_check(vkCreateImageView(device_, &vi, nullptr, &image_views_[i]), "swap image view");
    }

    VkImageCreateInfo msaa_info{};
    msaa_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
    msaa_info.imageType = VK_IMAGE_TYPE_2D;
    msaa_info.format = format_;
    msaa_info.extent = {extent_.width, extent_.height, 1};
    msaa_info.mipLevels = 1;
    msaa_info.arrayLayers = 1;
    msaa_info.samples = kMsaaSamples;
    msaa_info.tiling = VK_IMAGE_TILING_OPTIMAL;
    msaa_info.usage =
        VK_IMAGE_USAGE_TRANSIENT_ATTACHMENT_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
    msaa_info.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;

    VmaAllocator alloc = dev.allocator();
    VmaAllocationCreateInfo msaa_alloc_info{};
    msaa_alloc_info.usage = VMA_MEMORY_USAGE_GPU_ONLY;
    vk_check(vmaCreateImage(alloc, &msaa_info, &msaa_alloc_info, &msaa_image_, &msaa_alloc_,
                            nullptr),
             "msaa image");

    VkImageViewCreateInfo msaa_view_info{};
    msaa_view_info.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
    msaa_view_info.image = msaa_image_;
    msaa_view_info.viewType = VK_IMAGE_VIEW_TYPE_2D;
    msaa_view_info.format = format_;
    msaa_view_info.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
    msaa_view_info.subresourceRange.levelCount = 1;
    msaa_view_info.subresourceRange.layerCount = 1;
    vk_check(vkCreateImageView(device_, &msaa_view_info, nullptr, &msaa_view_), "msaa view");

  {
        VkAttachmentDescription color{};
        color.format = format_;
        color.samples = kMsaaSamples;
        color.loadOp = VK_ATTACHMENT_LOAD_OP_CLEAR;
        color.storeOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
        color.stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
        color.stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
        color.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
        color.finalLayout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;

        VkAttachmentDescription resolve{};
        resolve.format = format_;
        resolve.samples = VK_SAMPLE_COUNT_1_BIT;
        resolve.loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
        resolve.storeOp = VK_ATTACHMENT_STORE_OP_STORE;
        resolve.stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
        resolve.stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
        resolve.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
        resolve.finalLayout = VK_IMAGE_LAYOUT_PRESENT_SRC_KHR;

        VkAttachmentReference color_ref{};
        color_ref.attachment = 0;
        color_ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;

        VkAttachmentReference resolve_ref{};
        resolve_ref.attachment = 1;
        resolve_ref.layout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;

        VkSubpassDescription subpass{};
        subpass.pipelineBindPoint = VK_PIPELINE_BIND_POINT_GRAPHICS;
        subpass.colorAttachmentCount = 1;
        subpass.pColorAttachments = &color_ref;
        subpass.pResolveAttachments = &resolve_ref;

        VkSubpassDependency dep{};
        dep.srcSubpass = VK_SUBPASS_EXTERNAL;
        dep.dstSubpass = 0;
        dep.srcStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
        dep.dstStageMask = VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT;
        dep.dstAccessMask = VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT;

        VkAttachmentDescription attachments[] = {color, resolve};
        VkRenderPassCreateInfo rp_info{};
        rp_info.sType = VK_STRUCTURE_TYPE_RENDER_PASS_CREATE_INFO;
        rp_info.attachmentCount = 2;
        rp_info.pAttachments = attachments;
        rp_info.subpassCount = 1;
        rp_info.pSubpasses = &subpass;
        rp_info.dependencyCount = 1;
        rp_info.pDependencies = &dep;
        vk_check(vkCreateRenderPass(device_, &rp_info, nullptr, &render_pass_), "render pass");
    }

    framebuffers_.resize(image_views_.size());
    for (std::size_t i = 0; i < image_views_.size(); ++i) {
        VkImageView attachments[] = {msaa_view_, image_views_[i]};
        VkFramebufferCreateInfo fb{};
        fb.sType = VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO;
        fb.renderPass = render_pass_;
        fb.attachmentCount = 2;
        fb.pAttachments = attachments;
        fb.width = extent_.width;
        fb.height = extent_.height;
        fb.layers = 1;
        vk_check(vkCreateFramebuffer(device_, &fb, nullptr, &framebuffers_[i]), "framebuffer");
    }
    (void)instance;
}

void SwapchainContext::destroy(VulkanDevice& dev) {
    VkDevice d = dev.device();
    for (auto fb : framebuffers_) {
        vkDestroyFramebuffer(d, fb, nullptr);
    }
    framebuffers_.clear();
    if (render_pass_) {
        vkDestroyRenderPass(d, render_pass_, nullptr);
        render_pass_ = VK_NULL_HANDLE;
    }
    if (msaa_view_) {
        vkDestroyImageView(d, msaa_view_, nullptr);
        msaa_view_ = VK_NULL_HANDLE;
    }
    if (msaa_image_) {
        vmaDestroyImage(dev.allocator(), msaa_image_, msaa_alloc_);
        msaa_image_ = VK_NULL_HANDLE;
        msaa_alloc_ = VK_NULL_HANDLE;
    }
    for (auto view : image_views_) {
        vkDestroyImageView(d, view, nullptr);
    }
    image_views_.clear();
    if (swapchain_ && destroy_swapchain_) {
        destroy_swapchain_(d, swapchain_, nullptr);
        swapchain_ = VK_NULL_HANDLE;
    }
}

void SwapchainContext::acquire_next(VkDevice device, VkSemaphore sem, uint32_t& image_index) {
    vk_check(acquire_next_image_(device, swapchain_, UINT64_MAX, sem, VK_NULL_HANDLE, &image_index),
             "acquire next image");
}

void SwapchainContext::present(VkDevice device, VkQueue queue, VkSemaphore wait_sem,
                               uint32_t image_index) {
    VkPresentInfoKHR pi{};
    pi.sType = VK_STRUCTURE_TYPE_PRESENT_INFO_KHR;
    pi.waitSemaphoreCount = 1;
    pi.pWaitSemaphores = &wait_sem;
    pi.swapchainCount = 1;
    pi.pSwapchains = &swapchain_;
    pi.pImageIndices = &image_index;
    vk_check(queue_present_(queue, &pi), "queue present");
}

File: src/gpu/swapchain.rs
//! Swapchain、MSAA 图像与 RenderPass。

use ash::vk;

use super::device::VulkanDevice;
use super::instance::VulkanInstance;

/// MSAA 采样数。
pub const MSAA_SAMPLES: vk::SampleCountFlags = vk::SampleCountFlags::TYPE_4;

/// Swapchain 与帧缓冲相关资源。
pub struct SwapchainContext {
    /// Swapchain 扩展。
    pub swapchain_loader: ash::khr::swapchain::Device,
    /// Swapchain 句柄。
    pub swapchain: vk::SwapchainKHR,
    /// 表面格式。
    pub format: vk::SurfaceFormatKHR,
    ///  extent。
    pub extent: vk::Extent2D,
    /// 交换链图像。
    pub images: Vec<vk::Image>,
    /// 图像视图。
    pub image_views: Vec<vk::ImageView>,
    /// MSAA 颜色图。
    pub msaa_image: vk::Image,
    /// MSAA 视图。
    pub msaa_view: vk::ImageView,
    /// MSAA 分配（由 device 管理释放需在外部 free）。
    pub msaa_allocation: Option<gpu_allocator::vulkan::Allocation>, // take() on destroy
    /// RenderPass。
    pub render_pass: vk::RenderPass,
    /// 每 swapchain 图像的 framebuffer。
    pub framebuffers: Vec<vk::Framebuffer>,
}

impl SwapchainContext {
    /// 创建 swapchain、MSAA 与 render pass。
    pub fn new(
        inst: &VulkanInstance,
        dev: &mut VulkanDevice,
        pd: vk::PhysicalDevice,
        surface: vk::SurfaceKHR,
        width: u32,
        height: u32,
    ) -> Result<Self, String> {
        let caps = unsafe {
            inst.surface_loader
                .get_physical_device_surface_capabilities(pd, surface)
                .map_err(|e| format!("surface capabilities: {e}"))?
        };
        let formats = unsafe {
            inst.surface_loader
                .get_physical_device_surface_formats(pd, surface)
                .map_err(|e| format!("surface formats: {e}"))?
        };
        let format = formats
            .iter()
            .find(|f| {
                f.format == vk::Format::B8G8R8A8_SRGB
                    && f.color_space == vk::ColorSpaceKHR::SRGB_NONLINEAR
            })
            .copied()
            .unwrap_or(formats[0]);

        let extent = vk::Extent2D {
            width: width.clamp(caps.min_image_extent.width, caps.max_image_extent.width),
            height: height.clamp(caps.min_image_extent.height, caps.max_image_extent.height),
        };

        let image_count = (caps.min_image_count + 1).min(if caps.max_image_count > 0 {
            caps.max_image_count
        } else {
            u32::MAX
        });

        let queue_indices = [dev.families.graphics, dev.families.present];
        let create_info = vk::SwapchainCreateInfoKHR::default()
            .surface(surface)
            .min_image_count(image_count)
            .image_format(format.format)
            .image_color_space(format.color_space)
            .image_extent(extent)
            .image_array_layers(1)
            .image_usage(vk::ImageUsageFlags::COLOR_ATTACHMENT)
            .image_sharing_mode(vk::SharingMode::EXCLUSIVE)
            .pre_transform(caps.current_transform)
            .composite_alpha(vk::CompositeAlphaFlagsKHR::OPAQUE)
            .present_mode(vk::PresentModeKHR::FIFO)
            .clipped(true)
            .old_swapchain(vk::SwapchainKHR::null())
            .queue_family_indices(&queue_indices);

        let swapchain_loader = ash::khr::swapchain::Device::new(&inst.instance, &dev.device);
        let swapchain = unsafe { swapchain_loader.create_swapchain(&create_info, None) }
            .map_err(|e| format!("create_swapchain: {e}"))?;

        let images = unsafe { swapchain_loader.get_swapchain_images(swapchain) }
            .map_err(|e| format!("get_swapchain_images: {e}"))?;

        let image_views: Result<Vec<_>, _> = images
            .iter()
            .map(|img| {
                let view_info = vk::ImageViewCreateInfo::default()
                    .image(*img)
                    .view_type(vk::ImageViewType::TYPE_2D)
                    .format(format.format)
                    .subresource_range(vk::ImageSubresourceRange {
                        aspect_mask: vk::ImageAspectFlags::COLOR,
                        base_mip_level: 0,
                        level_count: 1,
                        base_array_layer: 0,
                        layer_count: 1,
                    });
                unsafe { dev.device.create_image_view(&view_info, None) }
                    .map_err(|e| format!("image_view: {e}"))
            })
            .collect();
        let image_views = image_views?;

        let (msaa_image, msaa_view, msaa_allocation) =
            create_msaa_image(dev, pd, extent, format.format)?;

        let render_pass = create_render_pass(&dev.device, format.format)?;

        let framebuffers: Result<Vec<_>, _> = image_views
            .iter()
            .map(|view| {
                let attachments = [msaa_view, *view];
                let fb_info = vk::FramebufferCreateInfo::default()
                    .render_pass(render_pass)
                    .attachments(&attachments)
                    .width(extent.width)
                    .height(extent.height)
                    .layers(1);
                unsafe { dev.device.create_framebuffer(&fb_info, None) }
                    .map_err(|e| format!("framebuffer: {e}"))
            })
            .collect();
        let framebuffers = framebuffers?;

        Ok(Self {
            swapchain_loader,
            swapchain,
            format,
            extent,
            images,
            image_views,
            msaa_image,
            msaa_view,
            msaa_allocation: Some(msaa_allocation),
            render_pass,
            framebuffers,
        })
    }

    /// 获取 framebuffer 数量。
    pub fn frame_count(&self) -> usize {
        self.framebuffers.len()
    }
}


/// 创建 MSAA 颜色附件。
fn create_msaa_image(
    dev: &mut VulkanDevice,
    pd: vk::PhysicalDevice,
    extent: vk::Extent2D,
    format: vk::Format,
) -> Result<(vk::Image, vk::ImageView, gpu_allocator::vulkan::Allocation), String> {
    let _ = pd;
    let img_info = vk::ImageCreateInfo::default()
        .image_type(vk::ImageType::TYPE_2D)
        .format(format)
        .extent(vk::Extent3D {
            width: extent.width,
            height: extent.height,
            depth: 1,
        })
        .mip_levels(1)
        .array_layers(1)
        .samples(MSAA_SAMPLES)
        .tiling(vk::ImageTiling::OPTIMAL)
        .usage(vk::ImageUsageFlags::TRANSIENT_ATTACHMENT | vk::ImageUsageFlags::COLOR_ATTACHMENT)
        .sharing_mode(vk::SharingMode::EXCLUSIVE)
        .initial_layout(vk::ImageLayout::UNDEFINED);

    let image = unsafe { dev.device.create_image(&img_info, None) }
        .map_err(|e| format!("msaa image: {e}"))?;
    let req = unsafe { dev.device.get_image_memory_requirements(image) };
    let allocation = dev
        .allocator
        .allocate(&gpu_allocator::vulkan::AllocationCreateDesc {
            name: "msaa",
            requirements: req,
            location: gpu_allocator::MemoryLocation::GpuOnly,
            linear: false,
            allocation_scheme: gpu_allocator::vulkan::AllocationScheme::GpuAllocatorManaged,
        })
        .map_err(|e| format!("msaa alloc: {e}"))?;
    unsafe {
        dev.device
            .bind_image_memory(image, allocation.memory(), allocation.offset())
            .map_err(|e| format!("bind msaa: {e}"))?;
    }
    let view_info = vk::ImageViewCreateInfo::default()
        .image(image)
        .view_type(vk::ImageViewType::TYPE_2D)
        .format(format)
        .subresource_range(vk::ImageSubresourceRange {
            aspect_mask: vk::ImageAspectFlags::COLOR,
            base_mip_level: 0,
            level_count: 1,
            base_array_layer: 0,
            layer_count: 1,
        });
    let view = unsafe { dev.device.create_image_view(&view_info, None) }
        .map_err(|e| format!("msaa view: {e}"))?;
    Ok((image, view, allocation))
}

/// 创建带 resolve 的 RenderPass。
fn create_render_pass(device: &ash::Device, format: vk::Format) -> Result<vk::RenderPass, String> {
    let color = vk::AttachmentDescription::default()
        .format(format)
        .samples(MSAA_SAMPLES)
        .load_op(vk::AttachmentLoadOp::CLEAR)
        .store_op(vk::AttachmentStoreOp::DONT_CARE)
        .stencil_load_op(vk::AttachmentLoadOp::DONT_CARE)
        .stencil_store_op(vk::AttachmentStoreOp::DONT_CARE)
        .initial_layout(vk::ImageLayout::UNDEFINED)
        .final_layout(vk::ImageLayout::COLOR_ATTACHMENT_OPTIMAL);

    let resolve = vk::AttachmentDescription::default()
        .format(format)
        .samples(vk::SampleCountFlags::TYPE_1)
        .load_op(vk::AttachmentLoadOp::DONT_CARE)
        .store_op(vk::AttachmentStoreOp::STORE)
        .stencil_load_op(vk::AttachmentLoadOp::DONT_CARE)
        .stencil_store_op(vk::AttachmentStoreOp::DONT_CARE)
        .initial_layout(vk::ImageLayout::UNDEFINED)
        .final_layout(vk::ImageLayout::PRESENT_SRC_KHR);

    let color_ref = vk::AttachmentReference::default()
        .attachment(0)
        .layout(vk::ImageLayout::COLOR_ATTACHMENT_OPTIMAL);
    let resolve_ref = vk::AttachmentReference::default()
        .attachment(1)
        .layout(vk::ImageLayout::COLOR_ATTACHMENT_OPTIMAL);

    let subpass = vk::SubpassDescription::default()
        .pipeline_bind_point(vk::PipelineBindPoint::GRAPHICS)
        .color_attachments(std::slice::from_ref(&color_ref))
        .resolve_attachments(std::slice::from_ref(&resolve_ref));

    let dependency = vk::SubpassDependency::default()
        .src_subpass(vk::SUBPASS_EXTERNAL)
        .dst_subpass(0)
        .src_stage_mask(vk::PipelineStageFlags::COLOR_ATTACHMENT_OUTPUT)
        .dst_stage_mask(vk::PipelineStageFlags::COLOR_ATTACHMENT_OUTPUT)
        .src_access_mask(vk::AccessFlags::empty())
        .dst_access_mask(vk::AccessFlags::COLOR_ATTACHMENT_WRITE);

    let attachments = [color, resolve];
    let rp_info = vk::RenderPassCreateInfo::default()
        .attachments(&attachments)
        .subpasses(std::slice::from_ref(&subpass))
        .dependencies(std::slice::from_ref(&dependency));

    unsafe { device.create_render_pass(&rp_info, None) }.map_err(|e| format!("render_pass: {e}"))
}
